### ISSUE TITLE
Using SparseAdam with LLaMA

### ISSUE THREAD

**Issue by Aradhye2002** [huggingface/transformers#30276]
Hi,

I'm working on PEFT techniques. I was trying out a line of work to sparsify gradients during backpropagation to reduce memory consumption. The model I'm using is llama. However I'm facing many issues:

1. SparseAdam, the optimizer for dense parameters but sparse gradients (my usecase), is not supported by the transformers.Trainer class. However I worked around this by tinkering with the source code and adding an option for SparseAdam.
2. nn.Embedding by default produces dense gradients, however for my usecase I require sparse gradients. Hence I need to set the sparse argument to True. I can't seem to find a high level argument to do this so I went ahead and changed the default value in my local pytorch nn.Embedding to True. This has some obvious issues.
3. But most importantly, even after 1 and 2, there seems to be something missing. It says in the error that "SparseAdam does not support dense gradients, please consider Adam instead" clearly implying that the gradients are still dense.

Can someone help me figure out 3? Solving 1 and 2 would be a bonus.
Thanks.

**Comment by amyeroberts** [comment]
cc @younesbelkada @pacman100
